3.0 update Cache in-memory

Rick-Anderson opened the pull request:
Fixes  #10479 Fixes #10218 Fixes #10858 Fixes #12634

[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.0&branch=pr-en-us-13812)

If there have been no changes in 3.0 to memory caching, then the review will be short.

Requires review: Does the  [System.Runtime.Caching/MemoryCache]( https://review.docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.0&branch=pr-en-us-13812#systemruntimecachingmemorycache) section apply to 3.0?

Changes to 3.0 version:

*  Removed Call `AddMemoryCache` in `ConfigureServices`: - That's not required. Anyone who needs to call `AddMemoryCache` in `ConfigureServices` will know how to do that.
* Added:
   If SizeLimit is not set, the cache grows without bound. The ASP.NET Core runtime does not trim the cache when system memory is low. Apps much be architected to:

    Limit cache growth.
    Call Compact or Remove when available memory is limited:
* Added section [MemoryCache.Compact](https://review.docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.0&branch=pr-en-us-13812#memorycachecompact)
* Ported all the code to 3.0 and tested. It all works fine. So the 3.0 version of this shows the 3.0 `StartUp`.

Rick-Anderson commented:
@mkArtakMSFT  given the 3.0 changes are trivial can you sign off on this PR or suggest a reviewer?

Changes to 3.0 version:

*  Removed Call `AddMemoryCache` in `ConfigureServices`: - That's not required. Anyone who needs to call `AddMemoryCache` in `ConfigureServices` will know how to do that.
* Ported all the code to 3.0 and tested. It all works fine. So the 3.0 version of this shows the 3.0 `StartUp`.

mkArtakMSFT commented:
@anurse can you assign somebody to review this?

BrennanConroy commented:
Could you explain what parts are new and need reviewing?

It looks like you split the docs into >=3.0 and <3.0 and a lot of the content was copied. So what is actually new?

Rick-Anderson commented:
@BrennanConroy  can you give me a couple days, I've decided to do some more work on it.

> So what is actually new?

See `Changes to 3.0 version:` at the top of the page.

Rick-Anderson commented:
@BrennanConroy  ready for review. I don't  think you need to review much unless there have been changes (beyond bug fixes) to caching for 3.0.

BrennanConroy commented:
>Removed Call AddMemoryCache in ConfigureServices: - That's not required. Anyone who needs to call AddMemoryCache in ConfigureServices will know how to do that.

I'm a bit confused by this. How will they know? We aren't telling them anymore.

BrennanConroy commented on aspnetcore/performance/caching/memory/3.0sample/RPcache/Services/MyMemoryCache.cs:
```diff
@@ -0,0 +1,19 @@
+    // using Microsoft.Extensions.Caching.Memory;
+    public class MyMemoryCache 
+    {
+        public MemoryCache Cache { get; set; }
```
Make the setter private

BrennanConroy commented on aspnetcore/performance/caching/memory.md:
```diff
@@ -9,25 +9,183 @@ uid: performance/caching/memory
+
+* When one cache entry is used to create another, the child copies the parent entry's expiration tokens and time-based expiration settings. The child isn't expired by manual removal or updating of the parent entry.
+
+* Use<xref:Microsoft.Extensions.Caching.Memory.ICacheEntry.PostEvictionCallbacks> to set the callbacks that will be fired after the cache entry is evicted from the cache.
```
```suggestion
* Use <xref:Microsoft.Extensions.Caching.Memory.ICacheEntry.PostEvictionCallbacks> to set the callbacks that will be fired after the cache entry is evicted from the cache.
```

Rick-Anderson commented:
@BrennanConroy  sorry I missed a couple issues tied to this doc.

* Select [MemoryCacheEntryOptions](https://review.docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.0&branch=pr-en-us-13812#memorycacheentryoptions) and review the two samples above, starting at *The following code gets or creates a cached item with absolute expiration:*
* Select [Additional notes](https://review.docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.0&branch=pr-en-us-13812#additional-notes) and review the first bullet, ie *Expiration doesn't happen in the background.*

If you like, I can add something about startup. Something like:

In-memory caching is a service that's referenced from your app using Dependency Injection. `Add the following:`  In `ConfigureServices`, the following API's call `AddMemoryCache` internally, so it's not necessary to call `AddMemoryCache`:

* `AddMvc`
* `AddMvcCore().AddRazorViewEngine`
* `AddResponseCaching`
* `AddControllersWithViews`
* `AddRazorPages`

BrennanConroy commented:
> the following API's call `AddMemoryCache` internally

Ah, didn't realize this. I still feel like we should tell users to add it explicitly if they want to use `IMemoryCache`.

I don't think we want to get into the business of maintaining a list of which middleware includes other middleware automatically.

Rick-Anderson commented:
>  I still feel like we should tell users to add it explicitly if they want to use `IMemoryCache`.

I'm not sure I understand.  Click on [Using IMemoryCache](https://review.docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-3.0&branch=pr-en-us-13812#using-imemorycache)
I'm using `IMemoryCache` without calling `AddMemoryCache`

BrennanConroy commented:
>I'm using IMemoryCache without calling AddMemoryCache

Yeah, because you're using one of the middleware that magically adds it behind the scenes.

Imagine you're a developer who wants to use `IMemoryCache` but doesn't have an app with MVC or Razor Pages or any of those middleware. You are going to read the document and try to inject `IMemoryCache` into one of your services and it will fail. And then you'll file an issue saying the doc doesn't explain how to add `IMemoryCache` to DI.

Side-note: Why is the warning in 2.2 under `Using IMemoryCache` no longer there in 3.0? I'm assuming it still applies.

Rick-Anderson commented:
> Imagine you're a developer who wants to use IMemoryCache but doesn't have an app with MVC or Razor Pages or any of those middleware.

I added the following to the **Additional Notes** section. I don't want to pollute the intro with this fairly clumsy information. Eilon often said, those .01% who want to do that will figure out how to do it. Not it's explicitly noted in the article. BTW, we had two issues created where customers complained they didn't need to call `AddMemoryCache`.

> Side-note: Why is the warning in 2.2 under `Using IMemoryCache` no longer there in 3.0?

Fixed. Thanks for catching that.

BrennanConroy commented:
Ok, I'm happy with the little blurb :+1:
